xapian-letor/tests/api_letor.cc

```diff
@@ -107,6 +107,8 @@ DEFINE_TESTCASE(preparetrainingfile, generated)
     string qrel = data_directory + "qrel.txt";
     Xapian::prepare_training_file(db_path, query, qrel, 10,
 				  "training_output.txt");
+
+    return true;
```

Comment:
Copy and pasting your implementation and comparing the output is obviously not a good way to write your tests - any bug in the implementation will be present in the copy and pasted version, and not be detected.  It would at least catch someone changing the implementation in a way which changes the output, but it's not a good approach.

But James does have a point - before this change, the output from `prepare_training_file()` was used as input to a test.  Now that output is totally ignored, and `prepare_training_file()` could produce literally any output and this test would pass.  That makes this a fairly poor test - we only ensure that it doesn't throw an exception, crash, or (if run under valgrind) leak memory or do something invalid that valgrind can detect.

Also writing a test not being easy isn't really a good reason not to write one.

We really want to ensure the output is **plausible** - testing the output can be read back in would at least catch bugs which result in output with an invalid format.  Comparing the output file with `testdata/training_data.txt` to ensure it's essentially the same would be better.  I don't think you can reliably just `diff` the files though, as that's probably too sensitive to rounding differences, but reading the two files together and checking corresponding real numbers only differ by a small amount should work.